aptos-move/framework/aptos-framework/sources/multisig_account.move

```diff
@@ -0,0 +1,1477 @@
+        emit_event(
+            &mut multisig_account_resource.execute_rejected_transaction_events,
+            ExecuteRejectedTransactionEvent {
+                transaction_id,
+                num_rejections,
+                executor: address_of(owner),
+            }
+        );
+    }
+
+    /// Called by the VM as part of transaction prologue, which is invoked during mempool transaction validation and as
+    /// the first step of transaction execution.
+    ///
+    /// Transaction payload is optional if it's already stored on chain for the transaction.
+    fun validate_multisig_transaction(
+        owner: &signer, multisig_account: address, payload: vector<u8>) acquires MultisigAccount {
+        assert_multisig_account_exists(multisig_account);
+        let multisig_account_resource = borrow_global<MultisigAccount>(multisig_account);
+        assert_is_owner(owner, multisig_account_resource);
+        let transaction_id = multisig_account_resource.last_executed_transaction_id + 1;
+        assert!(
+            table::contains(&multisig_account_resource.transactions, transaction_id),
+            error::invalid_argument(ETRANSACTION_NOT_FOUND),
+        );
+        let transaction = table::borrow(&multisig_account_resource.transactions, transaction_id);
+        let (num_approvals, _) = num_approvals_and_rejections(&multisig_account_resource.owners, transaction);
+        assert!(
+            num_approvals >= multisig_account_resource.signatures_required,
+            error::invalid_argument(ENOT_ENOUGH_APPROVALS),
+        );
+
+        // If the transaction payload is not stored on chain, verify that the provided payload matches the hashes stored
+        // on chain.
+        if (option::is_some(&transaction.payload_hash)) {
+            let payload_hash = option::borrow(&transaction.payload_hash);
+            assert!(
+                sha3_256(payload) == *payload_hash,
+                error::invalid_argument(EPAYLOAD_DOES_NOT_MATCH_HASH),
+            );
+        };
+    }
+
+    /// Post-execution cleanup for a successful multisig transaction execution.
+    /// This function is private so no other code can call this beside the VM itself as part of MultisigTransaction.
+    fun successful_transaction_execution_cleanup(
+        executor: address,
+        multisig_account: address,
+        transaction_payload: vector<u8>,
+    ) acquires MultisigAccount {
+        let multisig_account_resource = borrow_global_mut<MultisigAccount>(multisig_account);
+        let transaction_id = multisig_account_resource.last_executed_transaction_id + 1;
+        // Delete the transaction to keep storage efficient. Off-chain components can reconstruct the transaction and
+        // from events.
+        let transaction = table::remove(&mut multisig_account_resource.transactions, transaction_id);
+        let (num_approvals, _) = num_approvals_and_rejections(&multisig_account_resource.owners, &transaction);
```

Comment:
it seems this section is repeated in rejected/success|failed cleanup, probably worth a separate function

Reply:
This part?
        let transaction = table::remove(&mut multisig_account_resource.transactions, transaction_id);

I can make a separate function but that seems like overkill. And that function would still need to return the transaction object as failed cleanup needs it for events later

Reply:
the whole section from line 674 to 679